File: server_plugin/position-sender/src/main/java/org/blip/position/room/RoomManager.java
package org.blip.position.room;

import com.njkremer.Sqlite.DataConnectionException;
import com.njkremer.Sqlite.SqlStatement;

import org.blip.position.room.db.DBRoom;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.block.Block;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedQueue;

public class RoomManager {

    private ConcurrentLinkedQueue<Room> rooms = new ConcurrentLinkedQueue<>();

    public RoomManager() {
        try {
            List<DBRoom> rooms = SqlStatement.select(DBRoom.class).getList();

            for (DBRoom room : rooms) {
                Room memory = new Room(room.getUuid(),
                        room.getName(),
                        room.getWorld(),
                        room.getX(),
                        room.getY(),
                        room.getZ());

                World world = Bukkit.getWorld(UUID.fromString(room.getWorld()));

                if (null != world) {
                    manageBlocks(memory, world);
                    this.rooms.add(memory);
                }
            }
        } catch (DataConnectionException e) {
            e.printStackTrace();
        }
    }

    public Room create(String name, World world, int x, int y, int z) {
        Room room = new Room(name, world.getUID().toString(), x, y, z);
        DBRoom dbRoom = new DBRoom();
        dbRoom.setUuid(room.id());
        dbRoom.setName(name);
        dbRoom.setWorld(world.getUID().toString());
        dbRoom.setX(x);
        dbRoom.setY(y);
        dbRoom.setZ(z);
        try {
            SqlStatement.insert(dbRoom).execute();
        } catch (DataConnectionException e) {
            e.printStackTrace();
        }
        rooms.add(room);
        return room;
    }


    public ConcurrentLinkedQueue<Room> getRooms() {
        return rooms;
    }

    public void manageBlocks(Room room, World world) {
        SpreadGatherer spreadGatherer = new SpreadGatherer(world);
        List<Block> around = spreadGatherer.around(room.getOrigin().x,
                room.getOrigin().y,
                room.getOrigin().z);

        room.addBlocks(around);
    }
}

File: server_plugin/position-sender/src/main/kotlin/org/blip/position/room/RoomManager.kt
package org.blip.position.room

import org.blip.position.room.db.DBRoom
import org.bukkit.Bukkit
import org.bukkit.World
import org.ktorm.database.Database
import org.ktorm.database.use
import org.ktorm.dsl.forEach
import org.ktorm.dsl.from
import org.ktorm.dsl.insert
import org.ktorm.dsl.select
import org.ktorm.logging.Slf4jLoggerAdapter
import org.ktorm.support.sqlite.SQLiteDialect
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

class RoomManager {
    private val _rooms: ConcurrentLinkedQueue<Room>
    val database: Database

    val rooms: ConcurrentLinkedQueue<Room> get() = _rooms

    private val logger: Logger = LoggerFactory.getLogger(this.javaClass)

    init {
        database = Database.connect(
            url = "jdbc:sqlite:dolbyio.db",
            logger = Slf4jLoggerAdapter(logger),
            dialect = SQLiteDialect()
        )

        execSqlScript("db.sql", database)

        _rooms = ConcurrentLinkedQueue()
        database.from(DBRoom).select().forEach {

            val memory = Room(
                it[DBRoom.uuid],
                it[DBRoom.name],
                it[DBRoom.world],
                it[DBRoom.x]!!,
                it[DBRoom.y]!!,
                it[DBRoom.z]!!,
            )
            val world = Bukkit.getWorld(UUID.fromString(memory.world))
            if (null != world) {
                manageBlocks(memory, world)
                this.rooms.add(memory)
            }
        }
    }

    fun create(name: String?, world: World, x: Int, y: Int, z: Int): Room? {
        val room = Room(name, world.uid.toString(), x, y, z)

        database.insert(DBRoom) {
            set(it.uuid, room.id)
            set(it.name, room.name)
            set(it.world, world.uid.toString())
            set(it.x, x)
            set(it.y, y)
            set(it.z, z)
        }

        rooms.add(room)
        return room
    }

    fun manageBlocks(room: Room, world: World?) {
        val spreadGatherer = SpreadGatherer(world)
        val around = spreadGatherer.around(
            room.origin.x,
            room.origin.y,
            room.origin.z
        )
        room.addBlocks(around)
    }

    fun execSqlScript(filename: String, database: Database) {
        database.useConnection { conn ->
            conn.createStatement().use { statement ->
                javaClass.classLoader
                    ?.getResourceAsStream(filename)
                    ?.bufferedReader()
                    ?.use { reader ->
                        for (sql in reader.readText().split(';')) {
                            if (sql.any { it.isLetterOrDigit() }) {
                                statement.executeUpdate(sql)
                            }
                        }
                    }
            }
        }
    }
}